File: fcfs.cpp
#include<iostream>
using namespace std;

struct Process{
	int id;
	int wt=0;
	int at=0;
	int tot=0;
	int bt=0;
	int st=0;
};
int Time=0;
void make_same(Process source,Process &dest){
	dest.wt=source.wt;
	dest.at=source.at;
	dest.tot=source.tot;
	dest.bt=source.bt;
	dest.id=source.id;
}
void swap_processes(Process & p1,Process & p2){
	Process p3;
	make_same(p1,p3);
	make_same(p2,p1);
	make_same(p3,p2);
}
void sort_accto_arr_time(Process p[],int size){
	for(int i=0;i<size-1;i++){
		for(int j=0;j<size-1;j++){
			if(p[j].at>p[j+1].at){
				swap_processes(p[j],p[j+1]);
			}
		}
	}
}
void Draw_all_process_table(Process p[],int size){
	cout<<"\n\n";
	cout<<"+---------+"<<"------------+"<<"--------------+"<<"--------------+";
	cout<<endl;
	cout<<"| Process"<<" | "<<"Burst Time"<<" | "<<"Arrival Time"<<" | "<<"Waiting Time"<<" | "<<endl;
	cout<<"+---------+"<<"------------+"<<"--------------+"<<"--------------+";
	cout<<endl;

	for(int i=0;i<size;i++){
		cout<<"| P"<<p[i].id<<"      |";
		cout<<p[i].bt<<"           |";
		cout<<p[i].at<<"             |";
		cout<<p[i].wt<<"             |";
		cout<<endl;
	}
	cout<<"+---------+"<<"------------+"<<"--------------+"<<"--------------+";
	cout<<endl;
	cout<<"\n";
	
}
void draw_spaces(int bt){
	for(int i=0;i<bt;i++){
		cout<<" ";
	}
}
void draw_dashes(int bt){

	for(int j=0;j<bt;j++){
		cout<<"--";
	}
	cout<<" ";
	
}
void Draw_Gantt_Chart(Process p[],int size)
{
	cout<<"\n\nGantt Chart"<<endl;
	cout<<" ";
	for(int i=0;i<size;i++){
		draw_dashes(p[i].bt);
	}
	cout<<endl<<"|";

	for(int i=0;i<size;i++){
		draw_spaces(p[i].bt-1);
		cout<<"P"<<p[i].id;
		draw_spaces(p[i].bt-1);	
		cout<<"|";
	}

	cout<<endl;
	cout<<" ";
	for(int i=0;i<size;i++){
		draw_dashes(p[i].bt);
	}
	cout<<endl;

	for(int i=0;i<size;i++){

		cout<<Time;
		int offset=0;
		if(Time>9){
			offset=-1;
		
		}
		for(int j=0;j<p[i].bt+offset;j++){
			cout<<"  ";
		}
		Time+=p[i].bt;

	}
	cout<<Time<<"\n\n";
}
void calc_waiting_time(Process p[],int size){
	int t=0;
	for(int i=1;i<size;i++){
		p[i].wt=(p[i-1].wt+p[i-1].bt);
	}
	for(int i=0;i<size;i++){
		p[i].wt-=p[i].at;
	}
}

float calc_av_wt(Process p[],int size){
	float av_wt=0;
	for(int i=0;i<size;i++){
		av_wt+=p[i].wt;
	}
	return av_wt/size;
}
int main()
{
	Process p[5];


	p[0].at=2;
	p[1].at=5;
	p[2].at=1;
	p[3].at=0;
	p[4].at=4;

	p[0].bt=6;
	p[1].bt=2;
	p[2].bt=8;
	p[3].bt=3;
	p[4].bt=4;

	p[0].id=1;
	p[1].id=2;
	p[2].id=3;
	p[3].id=4;
	p[4].id=5;

	sort_accto_arr_time(p,5);
	p[0].wt=0;
	calc_waiting_time(p,5);

	cout<<"\n\nFirst Come First Serve (FCFS) Scheduling"<<endl;

	Draw_all_process_table(p,5);

	Draw_Gantt_Chart(p,5);

	cout<<"Average Waiting Time : "<<calc_av_wt(p,5)<<endl;

	


}

File: fcfs.py
import streamlit as st
import numpy as np
import pandas as pd

class Process():
	def __init__(self,wt=0,bt=0,at=0,pid=0,tat=0):
		self.bt=bt
		self.wt=wt 
		self.at=at 
		self.pid=pid
		self.tat=tat

	def __str__(self):
		return self.bt

def sort_acc_to_arrival():
	for i in range(0,len(st.session_state.p)-1):
		for j in range(0,len(st.session_state.p)-1-i):
			if st.session_state.p[j].at>st.session_state.p[j+1].at:
				p[j],p[j+1]=p[j+1],p[j]


def calc_waiting_time():
	t=0
	for i in range(1,len(st.session_state.p)):
		st.session_state.p[i].wt=st.session_state.p[i-1].wt+st.session_state.p[i-1].bt
	for i in range(0,len(st.session_state.p)):
		st.session_state.p[i].wt-=st.session_state.p[i].at

def calc_turnaround_time():
	for i in st.session_state.p:
		i.tat=i.wt+i.bt;

def get_dataframe(data,columns):
	return pd.DataFrame(data,columns=columns)

st.title("FCFS Scheduling")

p=[]
data=[]

if 'p' not in st.session_state:
    st.session_state.p = []
if 'data' not in st.session_state:
    st.session_state.data = []


pid=st.number_input("Enter Pid",step=1)
bt=st.number_input("Enter Burst Time",step=1)
at=st.number_input("Enter arrival Time",step=1)

add=st.button("Add process")
clr=st.button("Clear Table")

if clr:
	st.session_state.p = []
	st.session_state.data = []

if add:
	st.session_state.p.append(Process(pid=pid,bt=bt,at=at))


for i in range(len(st.session_state.p)):
	data.append([st.session_state.p[i].pid,st.session_state.p[i].bt,st.session_state.p[i].at])

df=get_dataframe(data,["Process",'Burst Time',"Arrival Time"])

st.header("Process Table")
st.dataframe(df)

if st.session_state.p!=[]:
	but=st.button("Calculate Waiting Time & Turnaround Time")


	if but:
		# sort_acc_to_arrival()
		# st.dataframe(df)

		calc_waiting_time()
		calc_turnaround_time()
		data=[]
		del df
		for i in range(len(st.session_state.p)):
			data.append([st.session_state.p[i].pid,st.session_state.p[i].wt,st.session_state.p[i].tat])

		df2=get_dataframe(data,["Process","Waiting Time","Turnaround Time"])
		st.dataframe(df2)

		st.header("Average waiting time")

		s=""
		av=0
		for i in st.session_state.p:
			av+=i.wt
			s+=str(i.wt)
			if i.pid<len(st.session_state.p):
				s+="+"
		av=av/len(st.session_state.p)
		st.latex(r"=\frac{"+s+"}{"+str(len(st.session_state.p))+"}="+str(av))

		st.header("Average Turnaround time")

		s=""
		av=0
		for i in st.session_state.p:
			av+=i.tat
			s+=str(i.tat)
			if i.pid<len(st.session_state.p):
				s+="+"
		av=av/len(st.session_state.p)
		st.latex(r"=\frac{"+s+"}{"+str(len(st.session_state.p))+"}="+str(av))


footer="""<style>


.footer {
position:fixed;
left: 0;
bottom: 0;
width: 100%;
background-color: black;
color: white;
text-align: center;
}
</style>
<div class="footer">
<p>Developed with ❤️ by Aditya Yadav</p>
</div>
"""
st.markdown(footer,unsafe_allow_html=True)